### ISSUE TITLE
unpaired broken with v1.15

### ISSUE THREAD

**Issue by tolot27** [broadinstitute/pilon#2]
Release 1.15 breaks the `--unpaired` funtionality. If I use PacBio CCS reads, pilon stops with the following exception:

```
# Attempting to fill gaps
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at com.simontuffs.onejar.Boot.run(Boot.java:340)
        at com.simontuffs.onejar.Boot.main(Boot.java:166)
Caused by: java.lang.IllegalStateException: Inappropriate call if not paired read
        at htsjdk.samtools.SAMRecord.requireReadPaired(SAMRecord.java:648)
        at htsjdk.samtools.SAMRecord.getFirstOfPairFlag(SAMRecord.java:706)
        at org.broadinstitute.pilon.BamFile$MateMap.addRead(BamFile.scala:219)
        at org.broadinstitute.pilon.BamFile$MateMap$$anonfun$addReads$1.apply(BamFile.scala:215)
        at org.broadinstitute.pilon.BamFile$MateMap$$anonfun$addReads$1.apply(BamFile.scala:215)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at org.broadinstitute.pilon.BamFile$MateMap.addReads(BamFile.scala:215)
        at org.broadinstitute.pilon.BamFile$MateMap.<init>(BamFile.scala:213)
        at org.broadinstitute.pilon.BamFile.recruitFlankReads(BamFile.scala:312)
        at org.broadinstitute.pilon.GapFiller$$anonfun$recruitReadsOfType$1.apply(GapFiller.scala:294)
        at org.broadinstitute.pilon.GapFiller$$anonfun$recruitReadsOfType$1.apply(GapFiller.scala:293)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at org.broadinstitute.pilon.GapFiller.recruitReadsOfType(GapFiller.scala:293)
        at org.broadinstitute.pilon.GapFiller.recruitUnpaired(GapFiller.scala:314)
        at org.broadinstitute.pilon.GapFiller.recruitLocalReads(GapFiller.scala:316)
        at org.broadinstitute.pilon.GapFiller.recruitReads(GapFiller.scala:318)
        at org.broadinstitute.pilon.GapFiller.assembleAcrossBreak(GapFiller.scala:51)
        at org.broadinstitute.pilon.GapFiller.fillGap(GapFiller.scala:40)
        at org.broadinstitute.pilon.GenomeRegion$$anonfun$identifyAndFixIssues$3.apply(GenomeRegion.scala:359)
        at org.broadinstitute.pilon.GenomeRegion$$anonfun$identifyAndFixIssues$3.apply(GenomeRegion.scala:357)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at org.broadinstitute.pilon.GenomeRegion.identifyAndFixIssues(GenomeRegion.scala:357)
        at org.broadinstitute.pilon.GenomeFile$$anonfun$processRegions$5.apply(GenomeFile.scala:116)
        at org.broadinstitute.pilon.GenomeFile$$anonfun$processRegions$5.apply(GenomeFile.scala:110)
        at scala.collection.Iterator$class.foreach(Iterator.scala:750)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
        at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:972)
        at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
        at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
        at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
        at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
        at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:969)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:152)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
        at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
With previous versions and the same dataset it works well.

**Comment by w1bw** [comment]
Thanks for the report. I'll look into it as soon as I can.

**Comment by w1bw** [comment]
Mathias,

I realize I don't have any test cases handy for --unpaired data. Can you please give this a try:

         s3.amazonaws.com/pilon/pilon-dev.jar

Please me know whether it fixes the problem.

**Comment by tolot27** [comment]
The dev version fixes the problem. Thanks!

**Comment by w1bw** [comment]
Great. I'll release it soon. Thanks again.

**Comment by kwongj** [comment]
Hi, thanks very much for your work on Pilon. Just wondering if this has become an issue again - I have been getting the same error recently with Pilon 1.20:
```
$ pilon --genome pre-pilon.fa --frags aln.sorted.bam --output pilon1 --fix all --mindepth 0.5 --changes --threads 16 --verbose
```
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.simontuffs.onejar.Boot.run(Boot.java:340)
	at com.simontuffs.onejar.Boot.main(Boot.java:166)
Caused by: java.lang.IllegalStateException: Inappropriate call if not paired read
	at htsjdk.samtools.SAMRecord.requireReadPaired(SAMRecord.java:648)
	at htsjdk.samtools.SAMRecord.getFirstOfPairFlag(SAMRecord.java:706)
	at org.broadinstitute.pilon.BamFile$MateMap.addRead(BamFile.scala:224)
	at org.broadinstitute.pilon.BamFile$MateMap$$anonfun$addReads$1.apply(BamFile.scala:220)
	at org.broadinstitute.pilon.BamFile$MateMap$$anonfun$addReads$1.apply(BamFile.scala:220)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.broadinstitute.pilon.BamFile$MateMap.addReads(BamFile.scala:220)
	at org.broadinstitute.pilon.BamFile$MateMap.<init>(BamFile.scala:218)
	at org.broadinstitute.pilon.BamFile.recruitFlankReads(BamFile.scala:339)
	at org.broadinstitute.pilon.GapFiller$$anonfun$recruitReadsOfType$1.apply(GapFiller.scala:367)
	at org.broadinstitute.pilon.GapFiller$$anonfun$recruitReadsOfType$1.apply(GapFiller.scala:366)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.broadinstitute.pilon.GapFiller.recruitReadsOfType(GapFiller.scala:366)
	at org.broadinstitute.pilon.GapFiller.recruitFrags(GapFiller.scala:375)
	at org.broadinstitute.pilon.GapFiller.recruitLocalReads(GapFiller.scala:389)
	at org.broadinstitute.pilon.GapFiller.recruitReads(GapFiller.scala:391)
	at org.broadinstitute.pilon.GapFiller.assembleAcrossBreak(GapFiller.scala:51)
	at org.broadinstitute.pilon.GapFiller.fixBreak(GapFiller.scala:45)
	at org.broadinstitute.pilon.GenomeRegion$$anonfun$identifyAndFixIssues$4.apply(GenomeRegion.scala:381)
	at org.broadinstitute.pilon.GenomeRegion$$anonfun$identifyAndFixIssues$4.apply(GenomeRegion.scala:379)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.broadinstitute.pilon.GenomeRegion.identifyAndFixIssues(GenomeRegion.scala:379)
	at org.broadinstitute.pilon.GenomeFile$$anonfun$processRegions$4.apply(GenomeFile.scala:120)
	at org.broadinstitute.pilon.GenomeFile$$anonfun$processRegions$4.apply(GenomeFile.scala:109)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:972)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:969)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:152)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
```
$ cat $(which pilon)
#!/bin/bash
exec java -mx16000m -ms120m -jar /home/linuxbrew/.linuxbrew/Cellar/pilon/1.20/pilon-1.20.jar "$@"
```
